File: crates/sargon-uniffi/src/signing/hd_signature_input.rs
use crate::prelude::*;
use paste::paste;

macro_rules! decl_hd_signature_input {
    (
        struct_name: $struct_name:ident,
        payload: $payload_id:ident
    ) => {
        #[derive(Clone, PartialEq, Eq, Hash, InternalConversion, uniffi::Record)]
        pub struct $struct_name {
            /// Hash which was signed.
            pub payload_id: $payload_id,

            /// The account or identity address of the entity which signed the hash,
            /// with expected public key and with derivation path to derive PrivateKey
            /// with.
            pub owned_factor_instance: OwnedFactorInstance,
        }
    };
    ($type:ty) => {
        paste! {
            use sargon::[< $type >] as [< Internal $type >];

            type [< InternalHDSignatureInputOf $type >] = sargon::HDSignatureInput<[< Internal $type >]>;

            decl_hd_signature_input!(
                struct_name: [< HDSignatureInputOf $type >],
                payload: [< $type >]
            );
        }
    };
}

decl_hd_signature_input!(TransactionIntentHash);
decl_hd_signature_input!(SubintentHash);
decl_hd_signature_input!(AuthIntentHash);

File: jvm/sargon-android/src/main/java/com/radixdlt/sargon/os/signing/HDSignatureInput.kt
package com.radixdlt.sargon.os.signing

import com.radixdlt.sargon.HdSignatureInputOfAuthIntentHash
import com.radixdlt.sargon.HdSignatureInputOfSubintentHash
import com.radixdlt.sargon.HdSignatureInputOfTransactionIntentHash
import com.radixdlt.sargon.HdSignatureOfTransactionIntentHash
import com.radixdlt.sargon.OwnedFactorInstance

data class HdSignatureInput<ID: Signable.ID> (
    /**
     * Hash which was signed.
     */
    val payloadId: ID,
    /**
     * The account or identity address of the entity which signed the hash,
     * with expected public key and with derivation path to derive PrivateKey
     * with.
     */
    val ownedFactorInstance: OwnedFactorInstance
)

fun HdSignatureInput<Signable.ID.Transaction>.intoSargon()
    = HdSignatureInputOfTransactionIntentHash(
        payloadId = payloadId.value,
        ownedFactorInstance = ownedFactorInstance
    )

fun HdSignatureInput<Signable.ID.Subintent>.intoSargon()
        = HdSignatureInputOfSubintentHash(
    payloadId = payloadId.value,
    ownedFactorInstance = ownedFactorInstance
)

fun HdSignatureInput<Signable.ID.Auth>.intoSargon()
        = HdSignatureInputOfAuthIntentHash(
    payloadId = payloadId.value,
    ownedFactorInstance = ownedFactorInstance
)